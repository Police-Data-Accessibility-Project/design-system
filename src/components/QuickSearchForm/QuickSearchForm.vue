<template>
	<FlexContainer class="p-0">
		<h2 class="mt-0">Search our database</h2>
		<p class="pb-4 md:pb-8">
			If you are looking for something specific, we may already be tracking it.
			<a href="https://pdap.io/data">Learn more about the data here.</a>
		</p>
	</FlexContainer>
	<FlexContainer
		alignment="center"
		class="pdap-quick-search-form h-full max-h-[75-vh] justify-start p-0"
	>
		<Form
			id="quick-search-form"
			class="flex flex-wrap gap-x-4"
			:error="error"
			:schema="formSchema"
			name="quickSearchForm"
			@change="handleChange"
			@submit="handleSubmit"
		>
			<Button
				type="submit"
				class="flex-grow-0 flex-shrink-0 basis-full max-w-[unset] mt-4"
				>Search Data Sources</Button
			>
		</Form>

		<p class="max-w-[unset] text-med">
			For example, you could search for
			<RouterLink v-if="hasSearchRoute" to="/search/stops/pittsburgh">
				stops in Pittsburgh
			</RouterLink>
			<a
				v-else
				target="_blank"
				rel="noreferrer"
				:href="`${props.baseUrlForRedirect}/search/stops/pittsburgh`"
			>
				stops in Pittsburgh
			</a>

			or
			<RouterLink v-if="hasSearchRoute" to="/search/complaints/all">
				complaints everywhere
			</RouterLink>
			<a
				v-else
				target="_blank"
				rel="noreferrer"
				:href="`${props.baseUrlForRedirect}/search/complaints/all`"
			>
				complaints everywhere
			</a>
			.
		</p>
	</FlexContainer>
</template>

<script setup lang="ts">
// Globals
import { useRouter } from 'vue-router';
// Components
import Button from '../Button/PdapButton.vue';
import FlexContainer from '../FlexContainer/FlexContainer.vue';
import Form from '../Form/PdapForm.vue';
import { RouterLink } from 'vue-router';

// Types
import { PdapInputTypes } from '../Input/types';
import { ref } from 'vue';

const router = useRouter();

const props = withDefaults(defineProps<{ baseUrlForRedirect: string }>(), {
	baseUrlForRedirect: 'https://data-sources.pdap.io',
});

const formSchema = [
	{
		id: 'search-term',
		name: 'searchTerm',
		label: 'What are you looking for?',
		type: PdapInputTypes.TEXT,
		placeholder: "Enter a keyword, or 'all'",
		value: '',
	},
	{
		id: 'location',
		name: 'location',
		label: 'From where?',
		type: PdapInputTypes.TEXT,
		placeholder: "Enter a place, or 'all'",
		value: '',
	},
];

const error = ref<string | undefined>(undefined);
const hasSearchRoute = router
	.getRoutes()
	.some((route) => route.path.includes('/search/'));

function handleChange(values: { location: string; searchTerm: string }) {
	// Reset error on form change
	if (error.value) {
		if (Object.values(values).some(Boolean)) {
			error.value = undefined;
		}
	}
}

function handleSubmit(values: { location: string; searchTerm: string }) {
	/**  Extra validation - backend expects 1 form value to be filled in.
	 *   The other input will be set to 'all'
	 */
	if (Object.values(values).every((val) => !val)) {
		error.value = 'Either a search term or a location is required.';
		return;
	}

	let { location, searchTerm } = values;

	location = location && location.length > 0 ? location : 'all';
	searchTerm = searchTerm && searchTerm.length > 0 ? searchTerm : 'all';

	// If search route exists, route to it
	if (hasSearchRoute) {
		router.push(`/search/${searchTerm}/${location}`);
	} else {
		window.location.assign(
			`${props.baseUrlForRedirect}/search/${searchTerm}/${location}`
		);
	}
}
</script>

<style>
@tailwind components;

@layer components {
	.pdap-quick-search-form .pdap-input {
		@apply flex-col flex-grow flex-shrink-0 basis-[45%] gap-y-0;
	}

	.pdap-quick-search-form .pdap-input-label {
		@apply justify-start;
	}
}
</style>
