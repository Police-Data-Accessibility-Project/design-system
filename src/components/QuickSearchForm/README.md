# QuickSearchForm
Renders a form to search the API `/search` endpoint with a search term and a location.

## Props

| name   | required? | types             | description                                                                  | default  |
| ------ | --------- | ----------------- | ---------------------------------------------------------------------------- | -------- |
| `mode` | no        | `'dev' \| 'prod'` | env —— controls which url users are sent to when form is rendered on pdap.io | `'prod'` |

## Notes

The different `mode` prop values result the following base url values in the eventual navigation 

| value    | base url                          |
| -------- | --------------------------------- |
| `'dev'`  | `'https://data-sources.pdap.dev'` |
| `'prod'` | `'https://data-sources.pdap.io'`  |

