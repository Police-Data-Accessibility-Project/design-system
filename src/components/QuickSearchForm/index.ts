import QuickSearchForm from './QuickSearchForm.vue';

export { QuickSearchForm };
