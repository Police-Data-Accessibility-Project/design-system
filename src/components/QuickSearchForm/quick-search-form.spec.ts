// Component
import QuickSearchForm from './QuickSearchForm.vue';

// Utils
import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, test, vi } from 'vitest';
import { nextTick } from 'vue';

// Mocks and stubs
const getRoutes = vi.fn(() => [{ path: '/search/:foo/:bar' }]);
const push = vi.fn();
vi.mock('vue-router', async () => {
	const actual: Record<string, unknown> = await vi.importActual('vue-router');
	return {
		...actual,
		useRouter: vi.fn(() => {
			return {
				push,
				getRoutes,
			};
		}),
	};
});

const assign = vi.fn();
vi.stubGlobal('location', {
	assign,
});

// Config
const base = {
	props: {
		mode: 'prod' as const,
	},
};

// Test
describe('QuickSearchForm component', () => {
	// Render
	test('Renders a QuickSearchForm', () => {
		const wrapper = mount(QuickSearchForm, base);

		expect(wrapper.find('.pdap-quick-search-form').exists()).toBe(true);
		expect(wrapper.html()).toMatchSnapshot();
	});

	describe('Form triggers router push when rendered in app with /search/ route', async () => {
		let searchInput;
		let locationInput;
		let wrapper;

		beforeEach(() => {
			getRoutes.mockReturnValueOnce([{ path: '/search/:foo/:bar' }]);

			wrapper = mount(QuickSearchForm, base);

			// Set input values
			searchInput = wrapper.find('#search-term');
			locationInput = wrapper.find('#location');
		});

		test('Form triggers router push with both values set', async () => {
			await searchInput.setValue('foo');
			await locationInput.setValue('bar');
			await nextTick();

			// Submit
			await wrapper.find('form').trigger('submit');
			await nextTick();
			await wrapper.vm.$forceUpdate();

			// Assert router push
			expect(push).toHaveBeenCalledWith('/search/foo/bar');
		});

		test('Form triggers router push with first value set, second unset', async () => {
			await searchInput.setValue('foo');
			await locationInput.setValue('');
			await nextTick();

			// Submit
			await wrapper.find('form').trigger('submit');
			await nextTick();
			await wrapper.vm.$forceUpdate();

			// Assert router push
			expect(push).toHaveBeenCalledWith('/search/foo/all');
		});

		test('Form triggers router push with second value set, first unset', async () => {
			await searchInput.setValue('');
			await locationInput.setValue('bar');
			await nextTick();

			// Submit
			await wrapper.find('form').trigger('submit');
			await nextTick();
			await wrapper.vm.$forceUpdate();

			// Assert router push
			expect(push).toHaveBeenCalledWith('/search/all/bar');
		});
	});

	describe('Form triggers window navigation when rendered in app without /search/ route — prod mode', async () => {
		let searchInput;
		let locationInput;
		let wrapper;

		beforeEach(() => {
			getRoutes.mockReturnValueOnce([
				{ path: '/' },
				{ path: '/foo' },
				{ path: '/bar' },
			]);

			wrapper = mount(QuickSearchForm, base);

			// Set input values
			searchInput = wrapper.find('#search-term');
			locationInput = wrapper.find('#location');
		});

		test('Form triggers navigation with both values set', async () => {
			await searchInput.setValue('foo');
			await locationInput.setValue('bar');
			await nextTick();

			// Submit
			await wrapper.find('form').trigger('submit');
			await nextTick();
			await wrapper.vm.$forceUpdate();

			expect(assign).toHaveBeenCalledWith(
				'https://data-sources.pdap.io/search/foo/bar'
			);
		});

		test('Form triggers navigation with first value set, second unset', async () => {
			await searchInput.setValue('foo');
			await locationInput.setValue('');
			await nextTick();

			// Submit
			await wrapper.find('form').trigger('submit');
			await nextTick();
			await wrapper.vm.$forceUpdate();

			expect(assign).toHaveBeenCalledWith(
				'https://data-sources.pdap.io/search/foo/all'
			);
		});

		test('Form triggers navigation with second value set, first unset', async () => {
			await searchInput.setValue('');
			await locationInput.setValue('bar');
			await nextTick();

			// Submit
			await wrapper.find('form').trigger('submit');
			await nextTick();
			await wrapper.vm.$forceUpdate();

			expect(assign).toHaveBeenCalledWith(
				'https://data-sources.pdap.io/search/all/bar'
			);
		});
	});

	test('Form triggers window navigation when rendered in app without /search/ route — custom baseUrlForRedirect', async () => {
		const baseUrlForRedirect = 'https://data-sources.pdap.dev';
		getRoutes.mockReturnValueOnce([
			{ path: '/' },
			{ path: '/foo' },
			{ path: '/bar' },
		]);

		const wrapper = mount(QuickSearchForm, {
			...base,
			props: {
				baseUrlForRedirect,
			},
		});

		// Set input values
		const searchInput = wrapper.find('#search-term');
		const locationInput = wrapper.find('#location');
		await searchInput.setValue('foo');
		await locationInput.setValue('bar');
		await nextTick();

		// Submit
		await wrapper.find('form').trigger('submit');
		await nextTick();
		await wrapper.vm.$forceUpdate();

		// Assert submit event
		expect(assign).toHaveBeenCalledWith(
			'https://data-sources.pdap.dev/search/foo/bar'
		);
	});

	test('Form errors when submitted with both inputs blank', async () => {
		const wrapper = mount(QuickSearchForm, base);

		// Submit
		await wrapper.find('form').trigger('submit.prevent');
		await nextTick();
		await wrapper.vm.$forceUpdate();

		// Assert error message
		expect(wrapper.find('.pdap-form-error-message').exists()).toBe(true);
		expect(wrapper.get('.pdap-form-error-message').text()).toBe(
			'Either a search term or a location is required.'
		);
	});
});
