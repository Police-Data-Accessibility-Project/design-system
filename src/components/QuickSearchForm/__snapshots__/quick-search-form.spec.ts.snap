// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QuickSearchForm component > Renders a QuickSearchForm 1`] = `
<div class="pdap-flex-container pdap-flex-container-start p-0">
  <h2 class="mt-0">Search our database</h2>
  <p class="pb-4 md:pb-8"> If you are looking for something specific, we may already be tracking it. <a href="https://pdap.io/data">Learn more about the data here.</a>
  </p>
</div>
<div class="pdap-flex-container pdap-flex-container-center pdap-quick-search-form h-full max-h-[75-vh] justify-start p-0">
  <form class="pdap-form flex flex-wrap gap-x-4" id="quick-search-form" name="quickSearchForm">
    <!--v-if-->
    <div class="pdap-grid-container pdap-input">
      <!-- Text -->
      <input id="search-term" name="searchTerm" placeholder="Enter a keyword, or &#x27;all&#x27;" type="text"><!-- Password -->
      <!--v-if-->
      <!--v-if-->
      <label class="pdap-input-label" for="search-term">What are you looking for?</label>
    </div>
    <div class="pdap-grid-container pdap-input">
      <!-- Text -->
      <input id="location" name="location" placeholder="Enter a place, or &#x27;all&#x27;" type="text"><!-- Password -->
      <!--v-if-->
      <!--v-if-->
      <label class="pdap-input-label" for="location">From where?</label>
    </div>
    <button class="pdap-button pdap-button-primary flex-grow-0 flex-shrink-0 basis-full max-w-[unset] mt-4" type="submit">Search Data Sources</button>
  </form>
  <p class="max-w-[unset] text-med"> For example, you could search for <a> stops in Pittsburgh </a> or <a> complaints everywhere </a> . </p>
</div>
`;
